src/SerialFirmata.cpp

```diff
@@ -296,27 +305,45 @@ void SerialFirmata::checkSerial()
         continue;
       }
 #endif
-      if (serialPort->available() > 0) {
-        Firmata.write(START_SYSEX);
-        Firmata.write(SERIAL_MESSAGE);
-        Firmata.write(SERIAL_REPLY | portId);
+      int availableBytes = serialPort->available();
+      if (availableBytes > 0) {
+        bool read = true;
 
-        if (bytesToRead == 0 || (serialPort->available() <= bytesToRead)) {
-          numBytesToRead = serialPort->available();
-        } else {
-          numBytesToRead = bytesToRead;
+#if defined(SERIAL_STORE_AND_FORWARD)
+        // check if reading should be delayed to collect some bytes before
+        // forwarding (for baud rates significantly below 57600 baud)
+        if (maxCharDelay[portId]) {
+          // inter character delay exceeded or more than 48 bytes available or more bytes available than required
+          read = (lastAvailableBytes[portId] > 0 && (currentMillis - lastReceive[portId]) >= maxCharDelay[portId])
+                 || (bytesToRead == 0 && availableBytes >= 48) || (bytesToRead > 0 && availableBytes >= bytesToRead);
+          if (availableBytes > lastAvailableBytes[portId]) {
```

Comment:
There is something wrong is the logic here, but I haven't had time to look into it far enough to figure out exactly what is wrong. The issue is that this conditional is only true a couple of times when my test application first runs. Therefore `read == true` forever after that.

My test is using an Arduino Mega with an Adafruit GPS shield (at 9600 baud) wired to RX1 on the Mega.